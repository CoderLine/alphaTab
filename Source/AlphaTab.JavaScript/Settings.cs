using AlphaTab.Collections;
using AlphaTab.Haxe.Js;
using AlphaTab.Util;
using Phase;

namespace AlphaTab
{
    public partial class CoreSettings
    {
        /// <summary>
        /// Gets or sets the script file url that will be used to spawn the workers.
        /// </summary>
        [JsonName("scriptFile")]
        public string ScriptFile { get; set; }

        /// <summary>
        /// Gets or sets the url to the fonts that will be used to generate the alphaTab font style.
        /// </summary>
        [JsonName("fontDirectory")]
        public string FontDirectory { get; set; }

        /// <summary>
        /// Gets or sets whether lazy loading for displayed elements is enabled.
        /// </summary>
        [JsonName("enableLazyLoading")]
        public bool EnableLazyLoading { get; set; } = true;

        public CoreSettings()
        {
            var global = Lib.Global;
            if (global.document && global.ALPHATAB_ROOT)
            {
                ScriptFile = global.ALPHATAB_ROOT;
                ScriptFile = EnsureFullUrl(ScriptFile);
                ScriptFile = AppendScriptName(ScriptFile);
            }
            else
            {
                ScriptFile = Environment.ScriptFile;
            }

            if (global.document && global.ALPHATAB_FONT)
            {
                FontDirectory = global.ALPHATAB_FONT;
                FontDirectory = EnsureFullUrl(FontDirectory);
            }
            else
            {
                FontDirectory = ScriptFile;
                if (!string.IsNullOrEmpty(FontDirectory))
                {
                    var lastSlash = FontDirectory.LastIndexOf('/');
                    if (lastSlash >= 0)
                    {
                        FontDirectory = FontDirectory.Substring(0, lastSlash) + "/Font/";
                    }
                }
            }
        }

        private static string EnsureFullUrl(string relativeUrl)
        {
            var global = Lib.Global;
            if (!relativeUrl.StartsWith("http") && !relativeUrl.StartsWith("https") && !relativeUrl.StartsWith("file"))
            {
                var root = new StringBuilder();
                root.Append(global.location.protocol);
                root.Append("//");
                if (global.location.hostname)
                {
                    root.Append(global.location.hostname);
                }

                if (global.location.port)
                {
                    root.Append(":");
                    root.Append(global.location.port);
                }

                // as it is not clearly defined how slashes are treated in the location object
                // better be safe than sorry here
                string directory = global.location.pathname.split("/").slice(0, -1).join("/");
                if (directory.Length > 0)
                {
                    if (!directory.StartsWith("/"))
                    {
                        root.Append("/");
                    }

                    root.Append(directory);
                }


                if (!relativeUrl.StartsWith("/"))
                {
                    root.Append("/");
                }

                root.Append(relativeUrl);

                return root.ToString();
            }

            return relativeUrl;
        }

        private static string AppendScriptName(string url)
        {
            // append script name
            if (!string.IsNullOrEmpty(url) && !url.EndsWith(".js"))
            {
                if (!url.EndsWith("/"))
                {
                    url += "/";
                }

                url += "AlphaTab.js";
            }

            return url;
        }

    }

    public partial class PlayerSettings
    {
        /// <summary>
        /// Gets or sets the URl of the sound font to be loaded.
        /// </summary>
        [JsonName("soundFont")]
        public string SoundFont { get; set; }

        /// <summary>
        /// Gets or sets the element that should be used for scrolling.
        /// </summary>
        [JsonName("scrollElement")]
        public string ScrollElement { get; set; } = "html,body";
    }

    /// <summary>
    /// This public class contains instance specific settings for alphaTab
    /// </summary>
    public partial class Settings
    {
        public static Settings FromJson(dynamic json)
        {
            return null; // dynamically implemented via macro
        }

        public void FillFromJson(dynamic json)
        {
            // dynamically implemented via macro
        }

        public static object ToJson(Settings settings)
        {
            return null; // dynamically implemented via macro
        }

        public void FillFromDataAttributes(dynamic dataAttributes)
        {
            object dataAttributesObject = dataAttributes;
            var keys = Platform.Platform.JsonKeys(dataAttributesObject);
            foreach (var key in keys)
            {
                SetProperty(key.ToLower(), dataAttributes[key]);
            }
        }

        public bool SetProperty(string property, dynamic value)
        {
            // dynamically implemented via macro
            return false;
        }

    }
}
