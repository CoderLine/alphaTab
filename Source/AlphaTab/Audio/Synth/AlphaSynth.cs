using System;
using AlphaTab.Audio.Synth.Bank;
using AlphaTab.Audio.Synth.Midi;
using AlphaTab.Audio.Synth.Synthesis;
using AlphaTab.Audio.Synth.Util;
using AlphaTab.IO;
using AlphaTab.Util;

namespace AlphaTab.Audio.Synth
{
    /// <summary>
    /// This is the main synthesizer component which can be used to
    /// play a <see cref="MidiFile"/> via a <see cref="ISynthOutput"/>.
    /// </summary>
    public class AlphaSynth : IAlphaSynth
    {
        private readonly MidiFileSequencer _sequencer;
        private readonly Synthesizer _synthesizer;

        private bool _isSoundFontLoaded;
        private bool _isMidiLoaded;
        private int _tickPosition;
        private double _timePosition;

        /// <summary>
        /// Gets the <see cref="ISynthOutput"/> used for playing the generated samples.
        /// </summary>
        public ISynthOutput Output { get; }

        /// <inheritdoc />
        public bool IsReady { get; private set; }

        /// <inheritdoc />
        public bool IsReadyForPlayback => IsReady && _isSoundFontLoaded && _isMidiLoaded;

        /// <inheritdoc />
        public PlayerState State { get; private set; }

        /// <inheritdoc />
        public LogLevel LogLevel
        {
            get => Logger.LogLevel;
            set => Logger.LogLevel = value;
        }

        /// <inheritdoc />
        public float MasterVolume
        {
            get => _synthesizer.MasterVolume;
            set
            {
                value = SynthHelper.ClampF(value, SynthConstants.MinVolume, SynthConstants.MaxVolume);
                _synthesizer.MasterVolume = value;
            }
        }

        /// <inheritdoc />
        public float MetronomeVolume
        {
            get => _synthesizer.MetronomeVolume;
            set
            {
                value = SynthHelper.ClampF(value, SynthConstants.MinVolume, SynthConstants.MaxVolume);
                _synthesizer.MetronomeVolume = value;
            }
        }

        /// <inheritdoc />
        public double PlaybackSpeed
        {
            get => _sequencer.PlaybackSpeed;
            set
            {
                value = SynthHelper.ClampD(value, SynthConstants.MinPlaybackSpeed, SynthConstants.MaxPlaybackSpeed);
                var oldSpeed = _sequencer.PlaybackSpeed;
                _sequencer.PlaybackSpeed = value;
                UpdateTimePosition(_timePosition * (oldSpeed / value));
            }
        }

        /// <inheritdoc />
        public int TickPosition
        {
            get => _tickPosition;
            set => TimePosition = _sequencer.TickPositionToTimePosition(value);
        }

        /// <inheritdoc />
        public double TimePosition
        {
            get => _timePosition;
            set
            {
                Logger.Debug("AlphaSynth", "Seeking to position " + value + "ms");

                // tell the sequencer to jump to the given position
                _sequencer.Seek(value);

                // update the internal position
                UpdateTimePosition(value);

                // tell the output to reset the already synthesized buffers and request data again
                Output.ResetSamples();
            }
        }

        /// <inheritdoc />
        public PlaybackRange PlaybackRange
        {
            get => _sequencer.PlaybackRange;
            set
            {
                _sequencer.PlaybackRange = value;
                if (value != null)
                {
                    TickPosition = value.StartTick;
                }
            }
        }

        /// <inheritdoc />
        public bool IsLooping
        {
            get => _sequencer.IsLooping;
            set => _sequencer.IsLooping = value;
        }

        /// <inheritdoc />
        public void Destroy()
        {
            Logger.Debug("AlphaSynth", "Destroying player");
            Stop();
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="AlphaSynth"/> class.
        /// </summary>
        /// <param name="output">The output to use for playing the generated samples.</param>
        public AlphaSynth(ISynthOutput output)
        {
            Logger.Debug("AlphaSynth", "Initializing player");
            State = PlayerState.Paused;

            Logger.Debug("AlphaSynth", "Creating output");
            Output = output;
            Output.Ready += () =>
            {
                IsReady = true;
                OnReady();
                CheckReadyForPlayback();
            };
            Output.Finished += () =>
            {
                // stop everything
                Stop();
                Logger.Debug("AlphaSynth", "Finished playback");
                OnFinished(new PlaybackFinishedEventArgs(_sequencer.IsLooping));
                if (_sequencer.IsLooping)
                {
                    Play();
                }
            };
            Output.SampleRequest += () =>
            {
                // synthesize buffer
                _sequencer.FillMidiEventQueue();
                _synthesizer.Synthesize();
                // send it to output
                Output.AddSamples(_synthesizer.SampleBuffer);
                // tell sequencer to check whether its work is done
                _sequencer.CheckForStop();
            };
            Output.SamplesPlayed += OnSamplesPlayed;

            Logger.Debug("AlphaSynth", "Creating synthesizer");
            _synthesizer = new Synthesizer(Output.SampleRate, SynthConstants.AudioChannels, 441, 3, 100);

            _sequencer = new MidiFileSequencer(_synthesizer);
            _sequencer.Finished += Output.SequencerFinished;

            Logger.Debug("AlphaSynth", "Opening output");
            Output.Open();
        }

        /// <inheritdoc />
        public bool Play()
        {
            if (State == PlayerState.Playing || !IsReadyForPlayback)
            {
                return false;
            }

            Output.Activate();

            Logger.Debug("AlphaSynth", "Starting playback");
            State = PlayerState.Playing;
            OnStateChanged(new PlayerStateChangedEventArgs(State, false));
            Output.Play();
            return true;
        }

        /// <inheritdoc />
        public void Pause()
        {
            if (State == PlayerState.Paused || !IsReadyForPlayback)
            {
                return;
            }

            Logger.Debug("AlphaSynth", "Pausing playback");
            State = PlayerState.Paused;
            OnStateChanged(new PlayerStateChangedEventArgs(State, false));
            Output.Pause();
            _synthesizer.NoteOffAll(false);
        }

        /// <inheritdoc />
        public void PlayPause()
        {
            if (State == PlayerState.Playing || !IsReadyForPlayback)
            {
                Pause();
            }
            else
            {
                Play();
            }
        }

        /// <inheritdoc />
        public void Stop()
        {
            if (State == PlayerState.Paused || !IsReadyForPlayback)
            {
                return;
            }

            Logger.Debug("AlphaSynth", "Stopping playback");
            State = PlayerState.Paused;
            Output.Pause();
            _synthesizer.NoteOffAll(true);
            TickPosition = _sequencer.PlaybackRange != null ? _sequencer.PlaybackRange.StartTick : 0;
            OnStateChanged(new PlayerStateChangedEventArgs(State, true));
        }

        /// <inheritdoc />
        public void LoadSoundFont(byte[] data)
        {
            Pause();

            var input = ByteBuffer.FromBuffer(data);
            try
            {
                Logger.Info("AlphaSynth", "Loading soundfont from bytes");
                var bank = new PatchBank();
                bank.LoadSf2(input);
                _synthesizer.LoadBank(bank);

                _isSoundFontLoaded = true;
                OnSoundFontLoaded();

                Logger.Info("AlphaSynth", "soundFont successfully loaded");
                CheckReadyForPlayback();
            }
            catch (Exception e)
            {
                Logger.Error("AlphaSynth", "Could not load soundfont from bytes " + e);
                OnSoundFontLoadFailed(e);
            }
        }

        private void CheckReadyForPlayback()
        {
            if (IsReadyForPlayback)
            {
                OnReadyForPlayback();
            }
        }

        /// <summary>
        /// Loads the given midi file for playback.
        /// </summary>
        /// <param name="midiFile">The midi file to load</param>
        // ReSharper disable once UnusedMember.Global
        public void LoadMidiFile(MidiFile midiFile)
        {
            Stop();

            try
            {
                Logger.Info("AlphaSynth", "Loading midi from model");

                _sequencer.LoadMidi(midiFile);
                _isMidiLoaded = true;
                OnMidiLoaded();
                Logger.Info("AlphaSynth", "Midi successfully loaded");
                CheckReadyForPlayback();

                TickPosition = 0;
            }
            catch (Exception e)
            {
                Logger.Error("AlphaSynth", "Could not load midi from model " + e);
                OnMidiLoadFailed(e);
            }
        }

        /// <inheritdoc />
        public void SetChannelMute(int channel, bool mute)
        {
            _synthesizer.SetChannelMute(channel, mute);
        }

        /// <inheritdoc />
        public void ResetChannelStates()
        {
            _synthesizer.ResetChannelStates();
        }

        /// <inheritdoc />
        public void SetChannelSolo(int channel, bool solo)
        {
            _synthesizer.SetChannelSolo(channel, solo);
        }

        /// <inheritdoc />
        public void SetChannelVolume(int channel, double volume)
        {
            volume = SynthHelper.ClampD(volume, SynthConstants.MinVolume, SynthConstants.MaxVolume);
            _synthesizer.SetChannelVolume(channel, volume);
        }

        /// <inheritdoc />
        public void SetChannelProgram(int channel, byte program)
        {
            program = SynthHelper.ClampB(program, SynthConstants.MinProgram, SynthConstants.MaxProgram);
            _sequencer.SetChannelProgram(channel, program);
            _synthesizer.SetChannelProgram(channel, program);
        }

        private void OnSamplesPlayed(int sampleCount)
        {
            var playedMillis = sampleCount / (double)_synthesizer.SampleRate * 1000;
            UpdateTimePosition(_timePosition + playedMillis);
        }

        private void UpdateTimePosition(double timePosition)
        {
            // update the real positions
            var currentTime = _timePosition = timePosition;
            var currentTick = _tickPosition = _sequencer.TimePositionToTickPosition(currentTime);

            var endTime = _sequencer.EndTime;
            var endTick = _sequencer.EndTick;

            Logger.Debug("AlphaSynth",
                "Position changed: (time: " + currentTime + "/" + endTime + ", tick: " + currentTick + "/" + endTime +
                ", Active Voices: " + _synthesizer.ActiveVoices + ", Free Voices: " + _synthesizer.FreeVoices + ")");
            OnPositionChanged(new PositionChangedEventArgs(currentTime, endTime, currentTick, endTick));
        }

        #region Events

        /// <inheritdoc />
        public event Action Ready;

        private void OnReady()
        {
            var handler = Ready;
            if (handler != null)
            {
                handler();
            }
        }

        /// <summary>
        /// Occurs when the playback of the whole midi file finished.
        /// </summary>
        public event Action<PlaybackFinishedEventArgs> Finished;

        private void OnFinished(PlaybackFinishedEventArgs e)
        {
            var handler = Finished;
            if (handler != null)
            {
                handler(e);
            }
        }

        /// <summary>
        /// Occurs when the playback state changes.
        /// </summary>
        public event Action<PlayerStateChangedEventArgs> StateChanged;

        private void OnStateChanged(PlayerStateChangedEventArgs e)
        {
            var handler = StateChanged;
            if (handler != null)
            {
                handler(e);
            }
        }

        /// <summary>
        /// Occurs when the soundfont was successfully loaded.
        /// </summary>
        public event Action SoundFontLoaded;

        private void OnSoundFontLoaded()
        {
            var handler = SoundFontLoaded;
            if (handler != null)
            {
                handler();
            }
        }

        /// <summary>
        /// Occurs when AlphaSynth is ready to start the playback.
        /// This is the case once the <see cref="ISynthOutput"/> is ready, a SoundFont was loaded and also a MidiFle is loaded.
        /// </summary>
        public event Action ReadyForPlayback;

        private void OnReadyForPlayback()
        {
            var handler = ReadyForPlayback;
            if (handler != null)
            {
                handler();
            }
        }

        /// <summary>
        /// Occurs when the soundfont failed to be loaded.
        /// </summary>
        public event Action<Exception> SoundFontLoadFailed;

        private void OnSoundFontLoadFailed(Exception e)
        {
            var handler = SoundFontLoadFailed;
            if (handler != null)
            {
                handler(e);
            }
        }

        /// <summary>
        /// Occurs when the midi file was successfully loaded.
        /// </summary>
        public event Action MidiLoaded;

        private void OnMidiLoaded()
        {
            var handler = MidiLoaded;
            if (handler != null)
            {
                handler();
            }
        }

        /// <summary>
        /// Occurs when the midi failed to be loaded.
        /// </summary>
        public event Action<Exception> MidiLoadFailed;

        private void OnMidiLoadFailed(Exception e)
        {
            var handler = MidiLoadFailed;
            if (handler != null)
            {
                handler(e);
            }
        }

        /// <summary>
        /// Occurs whenever the current time of the played audio changes.
        /// </summary>
        public event Action<PositionChangedEventArgs> PositionChanged;

        private void OnPositionChanged(PositionChangedEventArgs e)
        {
            var handler = PositionChanged;
            if (handler != null)
            {
                handler(e);
            }
        }

        #endregion
    }

    #region EventArgs

    /// <summary>
    /// Represents the progress of any data being loaded.
    /// </summary>
    public class ProgressEventArgs
    {
        /// <summary>
        /// Gets the currently loaded bytes.
        /// </summary>
        public int Loaded { get; }

        /// <summary>
        /// Gets the total number of bytes to load.
        /// </summary>
        public int Total { get; }

        /// <summary>
        /// Initializes a new instance of the <see cref="ProgressEventArgs"/> class.
        /// </summary>
        /// <param name="loaded"></param>
        /// <param name="total"></param>
        public ProgressEventArgs(int loaded, int total)
        {
            Loaded = loaded;
            Total = total;
        }
    }

    /// <summary>
    /// Represents the info when the playback of a whole song finished.
    /// </summary>
    public class PlaybackFinishedEventArgs
    {
        /// <summary>
        /// Gets a value indicating whether looping is active, and the playback will start automatically again at the beginning.
        /// </summary>
        public bool IsLooping { get; }

        /// <summary>
        /// Initializes a new instance of the <see cref="PlaybackFinishedEventArgs"/> class.
        /// </summary>
        /// <param name="isLooping"></param>
        public PlaybackFinishedEventArgs(bool isLooping)
        {
            IsLooping = isLooping;
        }
    }

    /// <summary>
    /// Represents the info when the player state changes.
    /// </summary>
    public class PlayerStateChangedEventArgs
    {
        /// <summary>
        /// The new state of the player.
        /// </summary>
        public PlayerState State { get; }

        /// <summary>
        /// Gets a value indicating whether the playback was stopped or only paused.
        /// </summary>
        /// <returns>true if the playback was stopped, false if the playback was started or paused</returns>
        public bool Stopped { get; }

        /// <summary>
        /// Initializes a new instance of the <see cref="PlayerStateChangedEventArgs"/> class.
        /// </summary>
        /// <param name="state">The state.</param>
        public PlayerStateChangedEventArgs(PlayerState state, bool stopped)
        {
            State = state;
            Stopped = stopped;
        }
    }

    /// <summary>
    /// Represents the info when the time in the synthesizer changes.
    /// </summary>
    public class PositionChangedEventArgs
    {
        /// <summary>
        /// Gets the current time in milliseconds.
        /// </summary>
        public double CurrentTime { get; }

        /// <summary>
        /// Gets the length of the played song in milliseconds.
        /// </summary>
        public double EndTime { get; }

        /// <summary>
        /// Gets the current time in midi ticks.
        /// </summary>
        public int CurrentTick { get; }

        /// <summary>
        /// Gets the length of the played song in midi ticks.
        /// </summary>
        public int EndTick { get; }

        /// <summary>
        /// Initializes a new instance of the <see cref="PositionChangedEventArgs"/> class.
        /// </summary>
        /// <param name="currentTime">The current time.</param>
        /// <param name="endTime">The end time.</param>
        /// <param name="currentTick">The current tick.</param>
        /// <param name="endTick">The end tick.</param>
        public PositionChangedEventArgs(double currentTime, double endTime, int currentTick, int endTick)
        {
            CurrentTime = currentTime;
            EndTime = endTime;
            CurrentTick = currentTick;
            EndTick = endTick;
        }
    }

    #endregion
}
