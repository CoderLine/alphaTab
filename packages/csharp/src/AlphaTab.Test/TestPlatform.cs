using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace AlphaTab;

static partial class TestPlatform
{
    public static readonly Lazy<string> AlphaTabProjectRoot = new(() =>
    {
        var currentDir = new DirectoryInfo(System.Environment.CurrentDirectory);
        while (currentDir != null)
        {
            if (currentDir.GetDirectories(".git").Length == 1)
            {
                return System.IO.Path.Join(currentDir.FullName, "packages", "alphatab");
            }

            currentDir = currentDir.Parent;
        }

        throw new IOException(
            $"Could not find repository root via working dir {System.Environment.CurrentDirectory}");
    });

    public static async Task<Uint8Array> LoadFile(string path)
    {
        await using var fs =
            new FileStream(Path.Combine(AlphaTabProjectRoot.Value, path), FileMode.Open);
        await using var ms = new MemoryStream();
        await fs.CopyToAsync(ms);
        return new Uint8Array(ms.ToArray());
    }

    public static async Task<T> LoadFileAsJson<T>(string path)
    {
        await using var fs =
            new FileStream(Path.Combine(AlphaTabProjectRoot.Value, path), FileMode.Open);
        return (await JsonSerializer.DeserializeAsync<T>(fs, JsonOptions))!;
    }

    private static readonly JsonSerializerOptions JsonOptions = new JsonSerializerOptions
    {
        PropertyNameCaseInsensitive = true,
        Converters = { new ArrayTupleConverterFactory() }
    };

    private class ArrayTupleConverterFactory  :JsonConverterFactory
    {
        public override bool CanConvert(Type typeToConvert)
        {
            if (!typeToConvert.IsGenericType)
            {
                return false;
            }

            if (typeToConvert.GetGenericTypeDefinition() != typeof(ArrayTuple<,>))
            {
                return false;
            }

            return true;
        }

        public override JsonConverter CreateConverter(
            Type type,
            JsonSerializerOptions options)
        {
            var typeArguments = type.GetGenericArguments();
            var keyType = typeArguments[0];
            var valueType = typeArguments[1];

            var converter = (JsonConverter)Activator.CreateInstance(
                typeof(ArrayTupleConverter<,>).MakeGenericType(keyType, valueType),
                BindingFlags.Instance | BindingFlags.Public,
                binder: null,
                args: new object[]{options},
                culture: null)!;

            return converter;
        }
    }

    private class ArrayTupleConverter<TKey, TValue> : JsonConverter<ArrayTuple<TKey, TValue>>
    {
        private readonly JsonConverter<TValue> _valueConverter;
        private readonly JsonConverter<TKey> _keyConverter;
        private readonly Type _keyType;
        private readonly Type _valueType;

        public ArrayTupleConverter(JsonSerializerOptions options)
        {
            _valueConverter = (JsonConverter<TValue>)options
                .GetConverter(typeof(TValue));

            _keyConverter = (JsonConverter<TKey>)options
                .GetConverter(typeof(TKey));

            _keyType = typeof(TKey);
            _valueType = typeof(TValue);
        }

        public override ArrayTuple<TKey, TValue> Read(
            ref Utf8JsonReader reader,
            Type typeToConvert,
            JsonSerializerOptions options)
        {
            if (reader.TokenType != JsonTokenType.StartArray)
            {
                throw new JsonException();
            }

            if (!reader.Read())
            {
                throw new JsonException();
            }
            var v0 = _keyConverter.Read(ref reader, _keyType, options)!;

            if (!reader.Read())
            {
                throw new JsonException();
            }

            var v1 = _valueConverter.Read(ref reader, _valueType, options)!;
            if (!reader.Read() || reader.TokenType != JsonTokenType.EndArray)
            {
                throw new JsonException();
            }

            return new ArrayTuple<TKey, TValue>(v0, v1);
        }

        public override void Write(Utf8JsonWriter writer, ArrayTuple<TKey, TValue> value, JsonSerializerOptions options)
        {
            writer.WriteStartArray();
            _keyConverter.Write(writer, value.V0, options);
            _valueConverter.Write(writer, value.V1, options);
            writer.WriteEndArray();
        }
    }

    public static Uint8Array LoadFileSync(string path)
    {
        using var fs =
            new FileStream(Path.Combine(AlphaTabProjectRoot.Value, path), FileMode.Open);
        using var ms = new MemoryStream();
        fs.CopyTo(ms);
        return new Uint8Array(ms.ToArray());
    }

    public static async Task SaveFile(string name, Uint8Array data)
    {
        var path = Path.Combine(AlphaTabProjectRoot.Value, name);
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        await using var fs = new FileStream(path, FileMode.Create);
        await fs.WriteAsync(data.Buffer.Raw, (int)data.ByteOffset, (int)data.Length);
    }

    public static Task<IList<string>> ListDirectory(string path)
    {
        return Task.FromResult((IList<string>)Directory
            .EnumerateFiles(Path.Combine(AlphaTabProjectRoot.Value, path))
            .Select(Path.GetFileName)
            .ToList());
    }

    public static string JoinPath(string path1, string path2, string path3)
    {
        return Path.Join(path1, path2, path3);
    }

    public static Task DeleteFile(string path)
    {
        if (File.Exists(path))
        {
            File.Delete(path);
        }

        return Task.CompletedTask;
    }

    public static T[] EnumValues<T>(Type type) where T : struct, Enum
    {
        return Enum.GetValues<T>();
    }

    public static IEnumerable<ArrayTuple<object?, object?>> MapAsUnknownIterable(object map)
    {
        if (map is IDictionary mapAsUnknownIterable)
        {
            foreach (DictionaryEntry v in mapAsUnknownIterable)
            {
                yield return new ArrayTuple<object?, object?>(v.Key, v.Value);
            }
        }
        else
        {
            throw new ArgumentException("Provided value was no map", nameof(map));
        }
    }

    public static IEnumerable<object?> TypedArrayAsUnknownIterable(object map)
    {
        if (map is IEnumerable enumerable)
        {
            foreach (var v in enumerable)
            {
                yield return v;
            }
        }
        else
        {
            throw new ArgumentException("Provided value was no map", nameof(map));
        }
    }

    public static IList TypedArrayAsUnknownArray(object? array)
    {
        return array switch
        {
            IList l => l,
            _ => throw new InvalidCastException(
                $"Unknown array type[{array?.GetType().FullName}]")
        };
    }
}
