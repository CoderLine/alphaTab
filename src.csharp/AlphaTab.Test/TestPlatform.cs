using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using AlphaTab.Core.EcmaScript;

namespace AlphaTab
{
    static partial class TestPlatform
    {
        public static async Task<Uint8Array> LoadFile(string path)
        {
            await using var fs = new FileStream(path, FileMode.Open);
            await using var ms = new MemoryStream();
            await fs.CopyToAsync(ms);
            return new Uint8Array(ms.ToArray());
        }

        public static async Task SaveFile(string name, Uint8Array data)
        {
            var path = Path.Combine("test-results", name);
            Directory.CreateDirectory(Path.GetDirectoryName(path));
            await using var fs = new FileStream(Path.Combine("test-results", name), FileMode.Create);
            await fs.WriteAsync(data.Data.Array!, data.Data.Offset, data.Data.Count);
        }

        public static Task<IList<string>> ListDirectory(string path)
        {
            return Task.FromResult((IList<string>) Directory.EnumerateFiles(path)
                .Select(Path.GetFileName)
                .ToList());
        }
    }
}
