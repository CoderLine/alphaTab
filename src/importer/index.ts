import { ScoreImporter } from '@src/importer/ScoreImporter';
import { ScoreLoader } from '@src/importer/ScoreLoader';
import { UnsupportedFormatError } from '@src/importer/UnsupportedFormatError';

export { ScoreImporter, ScoreLoader, UnsupportedFormatError };
