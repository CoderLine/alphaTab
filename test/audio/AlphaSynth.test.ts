import { AlphaSynthMidiFileHandler } from '@src/midi/AlphaSynthMidiFileHandler';
import { MidiFileGenerator } from '@src/midi/MidiFileGenerator';
import { MidiFile } from '@src/midi/MidiFile';
import { AlphaSynth } from '@src/synth/AlphaSynth';
import { AlphaTexImporter } from '@src/importer/AlphaTexImporter';
import { Score } from '@src/model/Score';
import { Settings } from '@src/Settings';
import { TestOutput } from '@test/audio/TestOutput';
import { TestPlatform } from '@test/TestPlatform';

describe('AlphaSynthTests', () => {
    it('pcm-generation', async () => {
        const data = await TestPlatform.loadFile('test-data/audio/default.sf2');
        let tex: string =
            '\\tempo 102 \\tuning E4 B3 G3 D3 A2 E2 \\instrument 25 . r.8 (0.4 0.3 ).8 ' +
            '(-.3 -.4 ).2 {d } | (0.4 0.3 ).8 r.8 (3.3 3.4 ).8 r.8 (5.4 5.3 ).4 r.8 (0.4 0.3 ).8 |' +
            ' r.8 (3.4 3.3 ).8 r.8 (6.3 6.4 ).8 (5.4 5.3 ).4 {d }r.8 |' +
            ' (0.4 0.3).8 r.8(3.4 3.3).8 r.8(5.4 5.3).4 r.8(3.4 3.3).8 | ' +
            'r.8(0.4 0.3).8(-.3 - .4).2 { d } | ';
        let importer: AlphaTexImporter = new AlphaTexImporter();
        importer.init(TestPlatform.createStringReader(tex), new Settings());
        let score: Score = importer.readScore();
        let midi: MidiFile = new MidiFile();
        let gen: MidiFileGenerator = new MidiFileGenerator(score, null, new AlphaSynthMidiFileHandler(midi));
        gen.generate();
        let testOutput: TestOutput = new TestOutput();
        let synth: AlphaSynth = new AlphaSynth(testOutput);
        synth.loadSoundFont(data, false);
        synth.loadMidiFile(midi);
        synth.play();
        let finished: boolean = false;
        synth.finished.on(() => {
            finished = true;
        });
        while (!finished) {
            testOutput.next();
        }
    });
});
