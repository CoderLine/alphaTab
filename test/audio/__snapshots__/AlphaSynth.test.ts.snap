// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlphaSynthTests midi-bank 1`] = `
Array [
  Map {
    "track" => 0,
    "tick" => 0,
    "type" => 176,
    "channel" => 0,
    "controller" => 0,
    "value" => 0,
  },
  Map {
    "track" => 0,
    "tick" => 0,
    "type" => 176,
    "channel" => 0,
    "controller" => 32,
    "value" => 77,
  },
  Map {
    "track" => 0,
    "tick" => 0,
    "type" => 176,
    "channel" => 1,
    "controller" => 0,
    "value" => 0,
  },
  Map {
    "track" => 0,
    "tick" => 0,
    "type" => 176,
    "channel" => 1,
    "controller" => 32,
    "value" => 77,
  },
  Map {
    "track" => 1,
    "tick" => 0,
    "type" => 176,
    "channel" => 2,
    "controller" => 0,
    "value" => 0,
  },
  Map {
    "track" => 1,
    "tick" => 0,
    "type" => 176,
    "channel" => 2,
    "controller" => 32,
    "value" => 50,
  },
  Map {
    "track" => 1,
    "tick" => 0,
    "type" => 176,
    "channel" => 3,
    "controller" => 0,
    "value" => 0,
  },
  Map {
    "track" => 1,
    "tick" => 0,
    "type" => 176,
    "channel" => 3,
    "controller" => 32,
    "value" => 50,
  },
  Map {
    "track" => 0,
    "tick" => 3840,
    "type" => 176,
    "channel" => 0,
    "controller" => 0,
    "value" => 7,
  },
  Map {
    "track" => 0,
    "tick" => 3840,
    "type" => 176,
    "channel" => 0,
    "controller" => 32,
    "value" => 104,
  },
  Map {
    "track" => 0,
    "tick" => 3840,
    "type" => 176,
    "channel" => 1,
    "controller" => 0,
    "value" => 7,
  },
  Map {
    "track" => 0,
    "tick" => 3840,
    "type" => 176,
    "channel" => 1,
    "controller" => 32,
    "value" => 104,
  },
  Map {
    "track" => 1,
    "tick" => 5760,
    "type" => 176,
    "channel" => 2,
    "controller" => 0,
    "value" => 31,
  },
  Map {
    "track" => 1,
    "tick" => 5760,
    "type" => 176,
    "channel" => 2,
    "controller" => 32,
    "value" => 32,
  },
  Map {
    "track" => 1,
    "tick" => 5760,
    "type" => 176,
    "channel" => 3,
    "controller" => 0,
    "value" => 31,
  },
  Map {
    "track" => 1,
    "tick" => 5760,
    "type" => 176,
    "channel" => 3,
    "controller" => 32,
    "value" => 32,
  },
]
`;
